lib/scheduler/thread_pool.rb

```diff
@@ -115,24 +129,40 @@ def thread_loop
         work = nil
 
         @mutex.synchronize do
-          @new_work.wait(@mutex, @idle_time)
+          # we may have already have work so no need
+          # to wait for signals, this also handles the race
+          # condition between spinning up threads and posting work
+          work = @queue.pop(timeout: 0)
+          @new_work.wait(@mutex, @idle_time) if !work
 
-          if @queue.empty?
+          if !work && @queue.empty?
             done = @threads.count > @min_threads
           else
-            work = @queue.pop
+            work ||= @queue.pop
 
             if work == :shutdown
               work = nil
               done = true
             end
           end
 
+          @busy_threads << Thread.current if work
+
+          if !done && work && @queue.length > 0 && @threads.length < @max_threads &&
+               @busy_threads.length == @threads.length
+            spawn_thread
```

Comment:
The parent of this thread will be that of another "worker" thread and I'm not completely sure how that might affection exceptions within a thread. Would it be safer and simpler to spin up this thread in `post` instead?

Reply:
too complicated to get the mechanics right, we would not be spinning up threads in time or have a giant lock on post that heavily slows down posting.